channeldb/payments.go

```diff
@@ -378,80 +387,81 @@ func fetchPayment(bucket kvdb.RBucket) (*MPPayment, error) {
 // fetchHtlcAttempts retrives all htlc attempts made for the payment found in
 // the given bucket.
 func fetchHtlcAttempts(bucket kvdb.RBucket) ([]HTLCAttempt, error) {
-	htlcs := make([]HTLCAttempt, 0)
+	htlcsMap := make(map[uint64]*HTLCAttempt)
 
-	err := bucket.ForEach(func(k, _ []byte) error {
-		aid := byteOrder.Uint64(k)
-		htlcBucket := bucket.NestedReadBucket(k)
+	err := bucket.ForEach(func(k, v []byte) error {
+		aid := byteOrder.Uint64(k[len(k)-8:])
 
-		attemptInfo, err := fetchHtlcAttemptInfo(
-			htlcBucket,
-		)
-		if err != nil {
-			return err
+		if _, ok := htlcsMap[aid]; !ok {
+			htlcsMap[aid] = &HTLCAttempt{}
 		}
-		attemptInfo.AttemptID = aid
 
-		htlc := HTLCAttempt{
-			HTLCAttemptInfo: *attemptInfo,
-		}
+		var err error
+		switch {
+		case bytes.HasPrefix(k, htlcAttemptInfoKey):
+			attemptInfo, err := readHtlcAttemptInfo(v)
+			if err != nil {
+				return err
+			}
 
-		// Settle info might be nil.
-		htlc.Settle, err = fetchHtlcSettleInfo(htlcBucket)
-		if err != nil {
-			return err
-		}
+			attemptInfo.AttemptID = aid
+			htlcsMap[aid].HTLCAttemptInfo = *attemptInfo
+
+		case bytes.HasPrefix(k, htlcSettleInfoKey):
+			htlcsMap[aid].Settle, err = readHtlcSettleInfo(v)
+			if err != nil {
+				return err
+			}
+
+		case bytes.HasPrefix(k, htlcFailInfoKey):
+			htlcsMap[aid].Failure, err = readHtlcFailInfo(v)
+			if err != nil {
+				return err
+			}
 
-		// Failure info might be nil.
-		htlc.Failure, err = fetchHtlcFailInfo(htlcBucket)
-		if err != nil {
-			return err
 		}
 
-		htlcs = append(htlcs, htlc)
 		return nil
 	})
 	if err != nil {
 		return nil, err
 	}
 
-	return htlcs, nil
-}
+	keys := make([]uint64, len(htlcsMap))
+	i := 0
+	for k := range htlcsMap {
+		keys[i] = k
+		i++
+	}
 
-// fetchHtlcAttemptInfo fetches the payment attempt info for this htlc from the
-// bucket.
-func fetchHtlcAttemptInfo(bucket kvdb.RBucket) (*HTLCAttemptInfo, error) {
-	b := bucket.Get(htlcAttemptInfoKey)
-	if b == nil {
-		return nil, errNoAttemptInfo
+	sort.Slice(keys, func(i, j int) bool {
```

Comment:
I think because of the btree structure of bbolt the buckets are always sorted. And because the attempt ID (the `keys` here) previously were buckets, they were automatically sorted. But yes, a comment might not hurt here.